utils/filesystem.c

```diff
@@ -0,0 +1,414 @@
+        return -1;
+
+    int src_fd = open(src, O_RDONLY, 0);
+    if (src < 0)
+        goto out;
+
+    pv_filesystem_file_copy_from_fd(src_fd, tmp_fd, true);
+    close_fd(&tmp_fd);
+
+    int ret = pv_filesystem_path_rename(tmp_path, dst);
+    if (ret < 0)
+        goto out;
+
+out:
+    if (src_fd > -1)
+        close_fd(&src_fd);
+
+    if (tmp_fd > -1)
+        close_fd(&tmp_fd);
+
+    if (pv_filesystem_path_exist(tmp_path))
+        pv_filesystem_path_remove(tmp_path, false);
+
+    pv_filesystem_path_sync(src);
+    pv_filesystem_path_sync(dst);
+
+    return ret;
+}
+
+size_t pv_filesystem_path_get_size(const char *path)
+{
+    struct stat st;
+
+    stat(path, &st);
+    return st.st_size;
+}
+
+int pv_filesystem_file_get_xattr(char *value, size_t size, const char *fname, const char *attr)
+{
+    ssize_t cur_size = getxattr(fname, attr, value, size);
+    return cur_size == size ? 0 : -1;
+}
+
+char *pv_filesystem_file_get_xattr_dup(const char *fname, const char *attr)
+{
+    ssize_t size = getxattr(fname, attr, NULL, 0);
+    if (size < 1)
+        return NULL;
+
+    char *val = malloc(size);
+    if (!val)
+        return NULL;
+
+    ssize_t cur_size = getxattr(fname, attr, val, size);
+
+    if (cur_size != size) {
+        free(val);
+        return NULL;
+    }
+
+    return val;
+}
+
+int pv_filesystem_file_set_xattr(const char *fname, const char *attr, const char *value)
+{
+    ssize_t size = getxattr(fname, attr, NULL, 0);
+
+    int flag = XATTR_REPLACE;
+    if (size < 0 && errno == ENODATA)
+        flag = XATTR_CREATE;
+
+    size = setxattr(fname, attr, value, strlen(value), flag);
+    return size > 0 ? 0 : -1;
+}
+
+ssize_t pv_filesystem_file_write_nointr(int fd, const char *buf, ssize_t size)
+{
+    ssize_t written = 0;
+
+    while (written != size) {
+        ssize_t cur_write = write(fd, buf + written, size - written);
+
+        if (cur_write < 0) {
+            if (errno == EINTR)
+                continue;
+            break;
+        }
+        written += cur_write;
+    }
+    return written;
+}
+
+ssize_t pv_filesystem_file_read_nointr(int fd, char *buf, ssize_t size)
+{
+    ssize_t total_read = 0;
+
+    while (total_read != size) {
+        int cur_read = read(fd, buf + total_read, size - total_read);
+
+        if (cur_read < 0) {
+            if (errno == EINTR)
+                continue;
+            break;
+        }
+        if (cur_read == 0)
+            break;
+        total_read += cur_read;
+    }
+    return total_read;
+}
+
+int pv_filesystem_file_lock(int fd)
+{
+    struct flock flock;
+
+    flock.l_whence = SEEK_SET;
+    /*Lock the whole file*/
+    flock.l_len = 0;
+    flock.l_start = 0;
+    flock.l_type = F_WRLCK;
+
+    int ret = -1;
+    while (ret < 0 && errno == EINTR)
+        ret = fcntl(fd, F_SETLK, &flock);
+
+    return ret;
+}
+
+int pv_filesystem_file_unlock(int fd)
+{
+    struct flock flock;
+
+    flock.l_whence = SEEK_SET;
+    /*Lock the whole file*/
```

Comment:
Yeah... I like the `//` too, this function has been copy-pasted from the old code so I think that the /* */ style was "the rule" it would be a pleasure to change it!

Reply:
Fied in e7aab3009a251e62a8f9a4eae6a162729ecd19b1